test/integration/compliance_history_test.go

```diff
@@ -777,8 +777,42 @@ var _ = Describe("GRC: [P1][Sev1][policy-grc] Test the compliance history API",
 		 }]`)
 		Expect(err).ToNot(HaveOccurred())
 
-		By("The Policy should be Compliant")
-		clusters = verifyPolicyOnAllClusters(ctx, policyNS, policyName, "Compliant", defaultTimeoutSeconds*2)
+		debugMsg := common.RegisterDebugMessage()
+
+		By("The Policy should be Compliant on all clusters")
+		Eventually(func(g Gomega) {
+			*debugMsg = ""
+
+			for _, cluster := range clusters {
+				replPol, err := clientHubDynamic.Resource(common.GvrPolicy).Namespace(cluster).Get(
+					ctx, policyNS+"."+policyName, metav1.GetOptions{},
+				)
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(replPol).NotTo(BeNil())
+
+				details, found, err := unstructured.NestedSlice(replPol.Object, "status", "details")
+				g.Expect(found).To(BeTrue())
+				g.Expect(err).NotTo(HaveOccurred())
+
+				det, ok := details[0].(map[string]interface{})
+				g.Expect(ok).To(BeTrue())
+
+				history, found, err := unstructured.NestedSlice(det, "history")
+				g.Expect(found).To(BeTrue())
+				g.Expect(err).NotTo(HaveOccurred())
+
+				item, ok := history[0].(map[string]interface{})
+				g.Expect(ok).To(BeTrue())
+
+				msg, found, err := unstructured.NestedString(item, "message")
+				g.Expect(found).To(BeTrue())
+				g.Expect(err).NotTo(HaveOccurred())
+
+				*debugMsg += fmt.Sprintf("Cluster %v current compliance message: %v\n", cluster, msg)
```

Comment:
Should `debugMsg` be reset to an empty string when successful?

Reply:
Alternatively, you could include the policy name in the message.

Reply:
The advantage of using `RegisterDebugMessage` is that the message is not printed anywhere if the test eventually passes, so it is not necessary in that case to reset the message. The downside is that you only get the final state of the message: it doesn't log information every time "inside" the Eventually loop.

The policy name is already logged by another piece here, so I don't think it's necessary.

Reply:
I get the advantage of `RegisterDebugMessage` but this PR calls it multiple times in a `Describe`, so if the last one fails, you get messages from the previous that were successful. Without context, it's confusing since you assume it's related to the failure.

Reply:
I'm assuming that when a `DeferCleanup` is put inside an `It`, it registers a cleanup function just for that `It`. Specifically, I'm assuming that the cleanup function is not called after later `It` functions in the same `Describe`.

But, the [ginkgo docs](https://onsi.github.io/ginkgo/#cleaning-up-our-cleanup-code-defercleanup) do not make this clear: reading them again, it sounds like it just makes a new `AfterEach` function, which I think *would* mean it gets run after other `It` functions in that `Describe`. Is that your interpretation?

As we know, docs are not always reliable. Observations are required. In [the most recent failure](https://storage.googleapis.com/test-platform-results/pr-logs/pull/stolostron_governance-policy-framework/788/pull-ci-stolostron-governance-policy-framework-main-test-e2e-grc-framework/1780595706227068928/build-log.txt), the "Creates a policy with a Gatekeeper constraint" test failed. The previous `It` in that `Describe` uses a `DeferCleanup` function that logs "Deleting the policy". That log only occurs once in the output: that function *was not* run again at the end of the gatekeeper test.

So, I think `DeferCleanup` is magic; you can use `RegisterDebugMessage` multiple times in a `Describe`, and only the messages for the relevant failing test should be printed.